### PR TITLE
streamingccl: add IngestionFrontierProcessor

### PR DESCRIPTION
**Pull request by adityamaru**
The IngestionFrontierProcessor (IFP) is fed rows from the IngestionProcessor.
Each row represents a partition and the latest resolved ts event
received for that partition. The IFP uses the existing util
`spanFrontier` (used by changefeeds) to keep track of all the rows
returned by the several ingestion procs, and resolves the minimum
resolved ts for a particular partition across all procs. This is then
written to the ingestion job progress indicating that all processors
have ingested a stream up to a particular resolved ts.

TODO: Job resumption from the resolved ts reported in its progress. Will do this in a follow up PR, along with a test.

Informs: #57400

Release note: None

### CODE REVIEW

**Comment by adityamaru** [comment]
Will rebase once the ingestion processor lands!

**Review by pbardea — COMMENTED** [review]
Generally LGTM -- I have one question about updating the job progress here vs in the job itself.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-:-MQmh63IEdruHnSttoWv:bkiqwhm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @pbardea)

___
*[pkg/ccl/streamingccl/stream_ingestion_frontier_processor.go, line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MQmdanA5e6q5_fsQSBR:-MQmdanB6By3FIAZSXh1:b-42klhl) ([raw file](https://github.com/cockroachdb/cockroach/blob/29f656b3e04f969153373884a9a7b49c7dda21d6/pkg/ccl/streamingccl/stream_ingestion_frontier_processor.go#L44)):*
> ```Go
> 	// jobProgressedFn, if non-nil, is called to checkpoint the stream ingestion
> 	// job's progress in the corresponding system job entry.
> 	jobProgressedFn func(context.Context, jobs.HighWaterProgressedFn) error
> ```

I think an alternative to this processor updating the job directly would be for this processor's Next() to return when there is an update to the frontier and then let the job itself be responsible for maintaining it's own progress. I think that letting the processor not need to worry about the existence of jobs might be a nice property so that this processor just takes in rows and outputs rows without side-effects. (Concretely, we would return the new frontier in `Next()` upon `frontierChanged`.)

I'm curious if you considered that approach/any tradeoffs that you see with it. I would imagine that there would be more overhead in that case, but I'm not sure how much.

I don't feel strongly either way but want to discuss the tradeoffs.
___
*[pkg/ccl/streamingccl/stream_ingestion_frontier_processor.go, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MQmg9KO5fYv5B79WSTf:-MQmg9KP6OHN_uFyKz1_:builkhk) ([raw file](https://github.com/cockroachdb/cockroach/blob/29f656b3e04f969153373884a9a7b49c7dda21d6/pkg/ccl/streamingccl/stream_ingestion_frontier_processor.go#L154)):*
> ```Go
> 	if !resolved.Timestamp.IsEmpty() && resolved.Timestamp.Less(sf.highWaterAtStart) {
> 		logcrash.ReportOrPanic(sf.Ctx, &sf.flowCtx.Cfg.Settings.SV,
> 			`got a resolved timestamp %s that is less than the initial high-water %s`,
> ```

nit: s/initial high-water/start time/ might be clearer here as an error message?

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58672)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pbardea)

___
*[pkg/ccl/streamingccl/stream_ingestion_frontier_processor.go, line 44 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MQmdanA5e6q5_fsQSBR:-MQwm-qV3t4h76iRvV4A:bh27t1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/29f656b3e04f969153373884a9a7b49c7dda21d6/pkg/ccl/streamingccl/stream_ingestion_frontier_processor.go#L44)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

I think an alternative to this processor updating the job directly would be for this processor's Next() to return when there is an update to the frontier and then let the job itself be responsible for maintaining it's own progress. I think that letting the processor not need to worry about the existence of jobs might be a nice property so that this processor just takes in rows and outputs rows without side-effects. (Concretely, we would return the new frontier in `Next()` upon `frontierChanged`.)

I'm curious if you considered that approach/any tradeoffs that you see with it. I would imagine that there would be more overhead in that case, but I'm not sure how much.

I don't feel strongly either way but want to discuss the tradeoffs.
</blockquote></details>

Changed to the above-described logic! I have tests for the frontier proc and the rows it returns. I will hook up the job progress and write tests for its progress update in a follow-up PR.

I also removed the LoadJob from the frontier processor which was being used to set the HighWaterAtStart field. Instead I plan to pass this in via the FrontierSpec once the job progress is hooked up.
___
*[pkg/ccl/streamingccl/stream_ingestion_frontier_processor.go, line 154 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MQmg9KO5fYv5B79WSTf:-MQnik0B8kPHiCA-Sicx:b-cypxrc) ([raw file](https://github.com/cockroachdb/cockroach/blob/29f656b3e04f969153373884a9a7b49c7dda21d6/pkg/ccl/streamingccl/stream_ingestion_frontier_processor.go#L154)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

nit: s/initial high-water/start time/ might be clearer here as an error message?
</blockquote></details>

updated.

**Comment by adityamaru** [comment]
@pbardea I can wait for your package refactor before merging this, but I think this should be ready for review. There are further tweaks as part of #59112 and a follow-up job progress PR, but I think this is the meatiest of the frontier logic.

**Review by pbardea — COMMENTED** [review]
This is looking good!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-:-MRQFlVT3WC_x3VeJ5G8:bjwdyz2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru and @pbardea)

___
*[pkg/ccl/streamingccl/stream_ingestion_frontier_processor_test.go, line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MRQFzbQ5VcydL1hrcRB:-MRQFzbQ5VcydL1hrcRC:b5elg0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8d05cad49f6f2e1b05adb38f4f0b306b98c6599/pkg/ccl/streamingccl/stream_ingestion_frontier_processor_test.go#L74)):*
> ```Go
> 		name                      string
> 		events                    partitionToEvent
> 		expectedFrontierTimestamp hlc.Timestamp
> ```

Maybe we can also add a test case exercising some error cases (maybe the case where the input sends a resolved timestamp before the start time?)
___
*[pkg/ccl/streamingccl/stream_ingestion_frontier_processor_test.go, line 145 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MRQFUhfCYvbDdlm5klO:-MRQFUhfCYvbDdlm5klP:btlnsg5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8d05cad49f6f2e1b05adb38f4f0b306b98c6599/pkg/ccl/streamingccl/stream_ingestion_frontier_processor_test.go#L145)):*
> ```Go
> 			sip, ok := proc.(*streamIngestionProcessor)
> 			if !ok {
> 				t.Fatal("expected the processor that's created to be a split and scatter processor")
> ```

s/split and scatter/stream ingestion
___
*[pkg/ccl/streamingccl/stream_ingestion_processor_planning.go, line 157 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MRQEHxUBFlPXYyqJ0AR:-MRQEHxUBFlPXYyqJ0AS:b-fgknr9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8d05cad49f6f2e1b05adb38f4f0b306b98c6599/pkg/ccl/streamingccl/stream_ingestion_processor_planning.go#L157)):*
> ```Go
> }
> 
> func (s *streamIngestionResultWriter) AddRow(ctx context.Context, row tree.Datums) error {
> ```

Can we add `AddRow implements the rowexec.rowResultWriter interface` comments here and on the methods below?
___
*[pkg/ccl/streamingccl/stream_ingestion_processor_planning.go, line 166 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MRQElwt9oZMG0JBgrFS:-MRQElwt9oZMG0JBgrFT:b-lr43bp) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8d05cad49f6f2e1b05adb38f4f0b306b98c6599/pkg/ccl/streamingccl/stream_ingestion_processor_planning.go#L166)):*
> ```Go
> 		// Decode the row and write the ts.
> 		var ingestedHighWatermark hlc.Timestamp
> 		if err := protoutil.Unmarshal([]byte(*row[0].(*tree.DBytes)),
> ```

Can we check that row is not empty and row[0] is non-nil and return an error to avoid a panic?
___
*[pkg/ccl/streamingccl/stream_ingestion_processor_test.go, line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MRQDP7E62yB1Xd2OACZ:-MRQDP7E62yB1Xd2OAC_:b-s8282r) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8d05cad49f6f2e1b05adb38f4f0b306b98c6599/pkg/ccl/streamingccl/stream_ingestion_processor_test.go#L58)):*
> ```Go
> 	var ok bool
> 	if events, ok = m.partitionEvents[address]; !ok {
> 		return nil, errors.Newf("No events found for paritition %s", address)
> ```

super nit (especially considering it's a test), but s/No/no :)
___
*[pkg/sql/execinfrapb/processors_bulk_io.proto, line 144 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MRQCsimARN86JhuBEQB:-MRQCsimARN86JhuBEQC:b2vlmf3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8d05cad49f6f2e1b05adb38f4f0b306b98c6599/pkg/sql/execinfrapb/processors_bulk_io.proto#L144)):*
> ```
>   // processor. It is used to convey the timestamp until which the ingestion
>   // job has ingested data.
> ```
Does this actually convey the timestamp up until which the ingestion job has ingested data? This is only the start time of the ingestion job right? Does it make sense to say that this is the timestamp at which the frontier is initialized to and the processor expects there to be no events passed to it at a lower timestamp?

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58672)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @pbardea)

___
*[pkg/ccl/streamingccl/stream_ingestion_frontier_processor_test.go, line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MRQFzbQ5VcydL1hrcRB:-MRRK2Ke7_TunSpJu9Wj:b-44dvan) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8d05cad49f6f2e1b05adb38f4f0b306b98c6599/pkg/ccl/streamingccl/stream_ingestion_frontier_processor_test.go#L74)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Maybe we can also add a test case exercising some error cases (maybe the case where the input sends a resolved timestamp before the start time?)
</blockquote></details>

added a test for the start time < checkpoint event case. The remaining error cases in the frontier processor arise from not being able to marshal/unmarshal the row returned by the ingestion processor.
___
*[pkg/ccl/streamingccl/stream_ingestion_frontier_processor_test.go, line 145 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MRQFUhfCYvbDdlm5klO:-MRQlS8g3O1GcWjCWQIX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8d05cad49f6f2e1b05adb38f4f0b306b98c6599/pkg/ccl/streamingccl/stream_ingestion_frontier_processor_test.go#L145)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

s/split and scatter/stream ingestion
</blockquote></details>

Done.
___
*[pkg/ccl/streamingccl/stream_ingestion_processor_planning.go, line 157 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MRQEHxUBFlPXYyqJ0AR:-MRQmsvI0q4L-8e-VkZD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8d05cad49f6f2e1b05adb38f4f0b306b98c6599/pkg/ccl/streamingccl/stream_ingestion_processor_planning.go#L157)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Can we add `AddRow implements the rowexec.rowResultWriter interface` comments here and on the methods below?
</blockquote></details>

Done.
___
*[pkg/ccl/streamingccl/stream_ingestion_processor_planning.go, line 166 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MRQElwt9oZMG0JBgrFS:-MRR58zSA8p8BoK682vS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8d05cad49f6f2e1b05adb38f4f0b306b98c6599/pkg/ccl/streamingccl/stream_ingestion_processor_planning.go#L166)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

Can we check that row is not empty and row[0] is non-nil and return an error to avoid a panic?
</blockquote></details>

Done.
___
*[pkg/ccl/streamingccl/stream_ingestion_processor_test.go, line 58 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MRQDP7E62yB1Xd2OACZ:-MRR5CZsEC6yWfbbKJen:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8d05cad49f6f2e1b05adb38f4f0b306b98c6599/pkg/ccl/streamingccl/stream_ingestion_processor_test.go#L58)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

super nit (especially considering it's a test), but s/No/no :)
</blockquote></details>

Done.
___
*[pkg/sql/execinfrapb/processors_bulk_io.proto, line 144 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MRQCsimARN86JhuBEQB:-MRR7IJR12VZJgjh4zX_:boea81s) ([raw file](https://github.com/cockroachdb/cockroach/blob/d8d05cad49f6f2e1b05adb38f4f0b306b98c6599/pkg/sql/execinfrapb/processors_bulk_io.proto#L144)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

> ```
>   // processor. It is used to convey the timestamp until which the ingestion
>   // job has ingested data.
> ```
Does this actually convey the timestamp up until which the ingestion job has ingested data? This is only the start time of the ingestion job right? Does it make sense to say that this is the timestamp at which the frontier is initialized to and the processor expects there to be no events passed to it at a lower timestamp?
</blockquote></details>

yup you're right. Updated!

**Review by pbardea — APPROVED** [review]
LGTM -- this should be good to merge after the rebase too once CI is happy with it.

**Comment by adityamaru** [comment]
There's a data race that I believe will be fixed by the refactor of the ingestion processor in #59139, so I'm going to wait on that getting checked in.

**Review by pbardea — APPROVED** [review]
:lgtm_strong: with a few nits -- nice!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-:-MS0BdU781r9kwww5MCK:bek2g9o)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @adityamaru and @pbardea)

___
*[pkg/ccl/streamingccl/streamingest/stream_ingestion_processor_planning.go, line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MS0BIuSD7_2USW_aX2j:-MS0BIuSD7_2USW_aX2k:bk5gjq4) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfdfa8b7d1cd9f4befa21918d1f287d6679bc29a/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor_planning.go#L31)):*
> ```Go
> 
> func init() {
> 	rowexec.NewStreamIngestionFrontierProcessor = newStreamIngestionFrontierProcessor
> ```

let's move this `init` to the frontier processor itself?
___
*[pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go, line 38 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MS06uxr1eav5EGVFPSq:-MS06uxr1eav5EGVFPSr:b-vzjw3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfdfa8b7d1cd9f4befa21918d1f287d6679bc29a/pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go#L38)):*
> ```Go
> 	a       rowenc.DatumAlloc
> 
> 	// input returns rows from one or more streamIngestion processors
> ```

nit: `.` at the end of the line
___
*[pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go, line 88 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MS09ohL4b10WKEvW1_k:-MS09ohM4ewWahjgW0Vk:bbcdkjh) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfdfa8b7d1cd9f4befa21918d1f287d6679bc29a/pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go#L88)):*
> ```Go
> 	sf.input.Start(ctx)
> 	ctx = sf.StartInternal(ctx, streamIngestionFrontierProcName)
> 	return ctx
> ```

nit: `sf.StartInternal(...` since we don't use ctx?
___
*[pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go, line 118 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MS090AaAfwir7vevMAt:-MS090AaAfwir7vevMAu:b-4yvrbf) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfdfa8b7d1cd9f4befa21918d1f287d6679bc29a/pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go#L118)):*
> ```Go
> 			if err != nil {
> 				sf.MoveToDraining(err)
> 				return nil, sf.DrainHelper()
> ```

nit: `break` for consistency instead?

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58672)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @pbardea)

___
*[pkg/ccl/streamingccl/streamingest/stream_ingestion_processor_planning.go, line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/58672#-MS0BIuSD7_2USW_aX2j:-MS38fxABVL_EHzd36D9:b-yoys1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfdfa8b7d1cd9f4befa21918d1f287d6679bc29a/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor_planning.go#L31)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

let's move this `init` to the frontier processor itself?
</blockquote></details>

ahh, weird how this got here. Done.

**Comment by adityamaru** [comment]
TFTR!

bors r=pbardea

**Pull request merged**
